File: data-struct-chatgpt/05.merge-intervals.js
/* Problem 5: Merge Intervals (backend-relevant: scheduling, bookings, time windows)

You’re given an array of intervals intervals, where intervals[i] = [start, end].
Merge all overlapping intervals and return an array of the non-overlapping intervals that cover all the intervals in the input.

Overlapping rule: [a,b] overlaps [c,d] if c <= b.

Examples:

[[1,3],[2,6],[8,10],[15,18]] → [[1,6],[8,10],[15,18]]

[[1,4],[4,5]] → [[1,5]] */

console.log(mergeIntervals([[1,3],[2,6],[8,10],[15,18]])) // [[1,6],[8,10],[15,18]]
console.log(mergeIntervals([[1,4],[4,5]]))               // [[1,5]]
console.log(mergeIntervals([[1,4],[0,2],[3,5]]))         // [[0,5]]
console.log(mergeIntervals([]))                          // []

function mergeIntervals(intervals){
    if(intervals.length === 0){return []}

    intervals.sort((a, b) => a[0] - b[0] || a[1] - b[1])
    console.log(intervals)
    const result = []
    for(const [a, b] of intervals){
        if(result.length === 0 || a > result[result.length - 1][1]){
            result.push([a, b])
        }else{
            result[result.length - 1][1] = Math.max(b, result[result.length - 1][1])
        }
    }
    return result
}

File: data-struct-chatgpt/05.merge-intervals.go
package main

import (
	"fmt"
	"sort"
)

func main() {
	fmt.Println("hello world")
	fmt.Println(mergeIntervals([][]int{{1, 3}, {2, 6}, {8, 10}, {15, 18}}))
	fmt.Println(mergeIntervals([][]int{{1, 4}, {4, 5}}))
	fmt.Println(mergeIntervals([][]int{{1, 4}, {0, 2}, {3, 5}}))
	fmt.Println(mergeIntervals([][]int{}))
}

func mergeIntervals(intervals [][]int) [][]int {
	// we should sort first
	if len(intervals) == 0 {
		return [][]int{}
	}

	sort.Slice(intervals, func(i, j int) bool {
		if intervals[i][0] == intervals[j][0] {
			return intervals[i][1] < intervals[j][1]
		}
		return intervals[i][0] < intervals[j][0]
	})

	result := make([][]int, 0, len(intervals))
	for i := 0; i < len(intervals); i++ {
		if len(result) == 0 || intervals[i][0] > result[len(result)-1][1] {
			result = append(result, []int{intervals[i][0], intervals[i][1]})
		} else {
			if intervals[i][1] > result[len(result)-1][1] {
				result[len(result)-1][1] = intervals[i][1]
			}
		}
	}

	return result
}
